safe_url_check/README.md

```diff
@@ -0,0 +1,43 @@
+`safe_url_check` for Dart
+=========================
+
+Utility to check if an untrusted URL is broken, without accidentally connecting
+to a private IP address.
+
+**Disclaimer:** This is not an officially supported Google product.
+
+When running in a Cloud environment a program usually has access to private IPv4
+addresses. This private IP-space might be used to grant access to database,
+caches, temporary credentials and various other services. If a program in such
+a cloud environment is checking untrusted URLs to see if a URL is broken, an
+attacker could fool the program into connecting to a private IP address by
+configuring DNS to resolve as such.
+
+This is generally undesirable. In most cases it is unlikely to cause any
+issues, as making a trivial `HEAD` or `GET` request to check if the URL is
+broken should be without side-effects. However, it's often preferable to harden
+security by protecting unauthorized access to the private IP space.
+
+This package offers a `safeUrlCheck` function, which makes a `HEAD` request and
+follows redirects after verifying that the host does not resolve to a private
+IPv4 address or locally unique IPv6 address.
+
+Note, it is plausible that it is desirable to restrict access to additional
+addresses space, pull-requests with suggestions are encouraged.
```

Comment:
Do we have a guide somewhere telling us about this as best practice?
I just wonder, are there other what kinds of urls we would not like to access?

Reply:
As this comment implies, there plausibly is other kinds of IPs that we don't want to access.

I reference RFCs above that justify blocking private IPv4 addresses. If you go reading through these RFCs you're likely to find more ip-blocks that could be blocked. I think loopback, link-local, and private ipv4 is the most sensitive ip-spaces, these are usually services on the same machine or on the same private virtual network (VPN).

Note: The risk of accessing these with a `GET` request should by definition be minimal, as `GET` is not allowed to have side effects.